src/signalrclient/connection_impl.cpp

```diff
@@ -85,173 +94,177 @@ namespace signalr
             }
 
             // there should not be any active transport at this point
-            _ASSERTE(!m_transport);
+            assert(!m_transport);
 
-            m_disconnect_cts = pplx::cancellation_token_source();
+            m_disconnect_cts = std::make_shared<cancellation_token>();
             m_start_completed_event.reset();
             m_connection_id = "";
         }
 
-        start_negotiate(m_base_url, 0)
-            .then([callback](pplx::task<void> prev_task)
+        start_negotiate(m_base_url, 0, [callback](std::exception_ptr exception)
         {
-            try
+            if (exception != nullptr)
             {
-                prev_task.get();
-                callback(nullptr);
+                callback(exception);
             }
-            catch (...)
+            else
             {
-                callback(std::current_exception());
+                callback(nullptr);
             }
         });
     }
 
-    pplx::task<void> connection_impl::start_negotiate(const std::string& url, int redirect_count)
+    void connection_impl::start_negotiate(const std::string& url, int redirect_count, std::function<void(std::exception_ptr)> callback)
     {
         if (redirect_count >= MAX_NEGOTIATE_REDIRECTS)
         {
-            return pplx::task_from_exception<void>(signalr_exception("Negotiate redirection limit exceeded."));
+            change_state(connection_state::disconnected);
+            callback(std::make_exception_ptr(signalr_exception("Negotiate redirection limit exceeded.")));
+            return;
         }
 
-        pplx::task_completion_event<void> start_tce;
-
         std::weak_ptr<connection_impl> weak_connection = shared_from_this();
-        auto token = m_disconnect_cts.get_token();
+        const auto& token = m_disconnect_cts;
 
-        pplx::task_from_result()
-            .then([weak_connection, url]()
-        {
-            auto connection = weak_connection.lock();
-            if (!connection)
+        negotiate::negotiate(*m_http_client, url, m_signalr_client_config,
```

Comment:
nit: Any way we can chunk this method into smaller methods would be nice for readability.